Essence of unknown in attribute overview and added history

LemonUnit123 opened the pull request:
- EssenceOfUnknown visible in Attribute-Tab
- EssenceOfUnknown gain and history-log after bossbattle
- EssenceOfUnknown lost and history-log after buying galactic secret

LemonUnit123 committed:
Zwischenstand WIP

Nach Bosskämpfen muss noch EssenceOfUnknown hinzugefügt werden + Eintrag in Historie

LemonUnit123 committed:
Merge branch 'master' into EssenceOfUnknown-in-Attribute-Overview-and-added-History

LemonUnit123 committed:
Final commit

- master gemerged
- WIP-logs und -tests entfernt
- essenceOfUnknown wird jetzt nach Bosskämpfen verdient und Historie geloggt

LemonUnit123 committed:
deleted wrong addEssenceGain

addEssenceGain deleted in main.js

LemonUnit123 committed:
Fix if 0 levels done after bossbattle

History log gets not written now, after bossbattle with 0 essenceOfUnknown gained

r-zander committed:
WIP Code Review: Essence of Unknow in Attributes tab

r-zander commented on js/classes.js:
```diff
@@ -725,6 +725,64 @@ class PointOfInterest extends Entity {
+ */
+
+// TODO not sure if necessary
+class EssenceOfUnknownHistory extends Entity {
```
That doesn't need to be an entity. Actually not even a class, a type definition and then plain objects should be enough

r-zander commented on js/classes.js:
```diff
@@ -725,6 +725,64 @@ class PointOfInterest extends Entity {
+     * @param {EssenceOfUnknownHistorySavedValues} savedValues
+     */
+    constructor(savedValues) {
+        super('Essence Of Unknown Entry', `Transaction: ${savedValues.type} ${savedValues.amount} from ${savedValues.source}`);
```
That description is never used, so can be dropped

r-zander commented on js/classes.js:
```diff
@@ -725,6 +725,64 @@ class PointOfInterest extends Entity {
+            target: JsTypes.String,
+            cycle: JsTypes.Number,
+            timestamp: JsTypes.Number,
+        }, this);
```
that's good and should probably be kept, even when dropping this class

r-zander commented on js/effects.js:
```diff
@@ -34,6 +34,10 @@ class EffectType {
 
+    // TODO why? and if so: gain | spend
+    static EssenceOfUnknown = new EffectType('+', attributes.essenceOfUnknown);
+    static EssenceOfUnknownNegative = new EffectType('-', attributes.essenceOfUnknown);
```
Current theory: Those are not necessary, if we customize the attribute balance entries similar to how Grid Strength works.

Introducing those EffectTypes has a lot of implications that I don't like, especially for a currency

r-zander commented on js/gameData.js:
```diff
@@ -201,6 +201,21 @@ class GameData {
+     * @property {string} source
+     * @property {string} [target]
+     * @property {number} cycle
+     * @property {number} timestamp
```
Might be collapsed with `EssenceOfUnknownHistorySavedValues`

r-zander commented on js/gameData.js:
```diff
@@ -210,7 +225,8 @@ class GameData {
      *     galacticSecrets: Object<GalacticSecretSavedValues>,
-     *     requirements: Object<RequirementSavedValues>
+     *     requirements: Object<RequirementSavedValues>,
+     *     essenceOfUnknownHistory: EssenceOfUnknownHistorySavedValues[]
```
yeah, not a fan of it. The general `gameData.savedValues` should be the inner values of entities but EssenceOfUnknown history entries are not really entities.

r-zander commented on js/main.js:
```diff
@@ -949,6 +949,11 @@ function createAttributeBalance(rowElement, effectTypes) {
+        else if (effectType === EffectType.EssenceOfUnknown) {
+            for(const entry of gameData.essenceOfUnknownHistory) {
+                createEssenceOfUnknownBalanceEntry(rowElement, entry);
+            }
```
yeah, that adds onto the weird handling :/ more and more convinced `createGridLoadBalance` is a better template to approach this.

r-zander commented on js/main.js:
```diff
@@ -985,6 +990,35 @@ function createAttributeBalance(rowElement, effectTypes) {
+            break;
+        case 'gain':
+            descriptionText = ('from ' + historyEntry.source + ', gained at ' + historyEntry.cycle + ' IC');
+            break;
```
Considering: target & source could be merged to "[related]Entity". And "lost at"/"gained FROM" could be decided inline and based on the sign of the entry.amount. Need to try this out, it would simply what needs to be saved, but might reduce clarity.

r-zander commented on js/main.js:
```diff
@@ -985,6 +990,35 @@ function createAttributeBalance(rowElement, effectTypes) {
+            descriptionText = ('from ' + historyEntry.source + ', gained at ' + historyEntry.cycle + ' IC');
+            break;
+    }
+    let actualEffect = historyEntry.type == 'lost' ? EffectType.EssenceOfUnknownNegative : EffectType.EssenceOfUnknown;
```
yeah, this is where it stretches too thin

r-zander commented on js/main.js:
```diff
@@ -1110,6 +1144,12 @@ function createAttributesUI() {
 
+    // EssenceOfUnknown
+    const essenceOfUnknownRow = createAttributeRow(attributes.essenceOfUnknown);
+    Dom.get(essenceOfUnknownRow).byClass('balance').classList.remove('hidden');
+    createAttributeBalance(essenceOfUnknownRow, [EffectType.EssenceOfUnknown, EffectType.EssenceOfUnknownFactor]);
+    rows.push(essenceOfUnknownRow);
+
```
That on the other hand is rather smooth. But for me doesn't justify all the quirks "under the hood"

r-zander commented on js/main.js:
```diff
@@ -1982,6 +2022,7 @@ function updateStationOverview() {
 
     const essenceOfUnknown = attributes.essenceOfUnknown.getValue();
     formatValue(Dom.get().byId('essenceOfUnknownDisplay'), essenceOfUnknown, {forceInteger: true, keepNumber: true});
+    formatValue(Dom.get().bySelector('#attributeRows > .essenceOfUnknown > .value > data'), essenceOfUnknown);
```
that should also be a `{forceInteger: true, keepNumber: true}` I guess

r-zander commented on js/main.js:
```diff
@@ -2072,6 +2113,49 @@ function progressGalacticSecrets() {
+    const month = String(date.getMonth() + 1).padStart(2, '0'); // months are 0-based
+    const year = date.getFullYear();
+
+    const formattedDate = `${day}.${month}.${year}`;
```
that's a bit wild. Just saving the timestamp should be fully sufficient and should we ever decide to display the date, we can still format at this point.

r-zander commented on js/main.js:
```diff
@@ -2072,6 +2113,49 @@ function progressGalacticSecrets() {
+    // const slot = Dom.get().byId('attributeRows');
+    // const rowElement = Dom.get(slot).byClass('essenceOfUnknown');
+    
+    if (!gameData.essenceOfUnknownHistory) {
+        gameData.essenceOfUnknownHistory = [];
+    }
```
That should not be a thing - could be an assert though.

r-zander commented on index.html:
```diff
@@ -584,7 +584,7 @@ <h3>Settings</h3>
                             </tr>
                         </template>
-                        <template id="balanceEntryTemplate">
+                        <template id="balanceEntryTemplate" class="keep">
```
Need to check if the EssenceOfUnknownBalanceEntry could use its own template. Generally not the biggest fan of keeping templates around, but seems to be necessary here as the on compile time or even game start it is not foreseeable how many entries are required.

r-zander committed:
Merge branch 'refs/heads/master' into EssenceOfUnknown-in-Attribute-Overview-and-added-History

r-zander committed:
Refactored essence of unknown history.

- Moved from EffectType logic with SavedValues to a more custom approach for balance entries
- Moved from pre-rendered and saved entries, to referential saved entries that can be rendered on runtime
